File: space_invaders.py
import pygame
import random
import math

# Initialize Pygame
pygame.init()

# Screen dimensions
WIDTH = 800
HEIGHT = 600

# Colors
WHITE = (255, 255, 255)
RED = (255, 0, 0)
GREEN = (0, 255, 0)

# Create the screen
screen = pygame.display.set_mode((WIDTH, HEIGHT))
pygame.display.set_caption("Space Invaders")

# Load spaceship image
spaceship_img = pygame.image.load("spaceship.png")
spaceship_img = pygame.transform.scale(spaceship_img, (20, 60))

# Load enemy image
enemy_img = pygame.image.load("enemy.png")
enemy_img = pygame.transform.scale(enemy_img, (60, 60))

# Load projectile image
projectile_img = pygame.image.load("projectile.png")
projectile_img = pygame.transform.scale(projectile_img, (8, 8))

# Load enemy projectile image
enemy_projectile_img = pygame.image.load("enemy_projectile.png")
enemy_projectile_img = pygame.transform.scale(enemy_projectile_img, (8, 8))

# Spaceship class
class Spaceship:
    def __init__(self):
        self.x = WIDTH // 2
        self.xdir = 0

    def show(self):
        screen.blit(spaceship_img, (self.x, HEIGHT - 20))

    def move(self):
        self.x += self.xdir * 5

    def set_dir(self, dir):
        self.xdir = dir

# Enemy class
class Enemy:
    def __init__(self, x, y):
        self.x = x
        self.y = y
        self.xdir = 1
        self.speed = 2
        self.amplitude = 50
        self.frequency = 0.02
        self.offset = random.uniform(0, 100)

    def show(self):
        screen.blit(enemy_img, (self.x, self.y))

    def move(self):
        self.x += self.xdir * self.speed
        self.y += math.sin((pygame.time.get_ticks() + self.offset) * self.frequency) * self.amplitude * self.frequency
        if self.x > WIDTH or self.x < 0:
            self.xdir *= -1
            self.y += 20

# Projectile class
class Projectile:
    def __init__(self, x, y, is_enemy=False):
        self.x = x
        self.y = y
        self.to_delete = False
        self.is_enemy = is_enemy

    def show(self):
        if self.is_enemy:
            screen.blit(enemy_projectile_img, (self.x, self.y))
        else:
            screen.blit(projectile_img, (self.x, self.y))

    def move(self):
        self.y += 5 if self.is_enemy else -5

    def hits(self, target):
        d = math.sqrt((self.x - target.x) ** 2 + (self.y - target.y) ** 2)
        return d < 30

    def destroy(self):
        self.to_delete = True

# Initialize game objects
spaceship = Spaceship()
enemies = [Enemy(i * 80 + 80, 60) for i in range(10)]
projectiles = []
enemy_projectiles = []

# Game loop
running = True
while running:
    screen.fill((0, 0, 0))

    for event in pygame.event.get():
        if event.type == pygame.QUIT:
            running = False

        if event.type == pygame.KEYDOWN:
            if event.key == pygame.K_SPACE:
                projectiles.append(Projectile(spaceship.x, HEIGHT - 20))
            elif event.key == pygame.K_RIGHT:
                spaceship.set_dir(1)
            elif event.key == pygame.K_LEFT:
                spaceship.set_dir(-1)

        if event.type == pygame.KEYUP:
            if event.key in (pygame.K_RIGHT, pygame.K_LEFT):
                spaceship.set_dir(0)

    spaceship.show()
    spaceship.move()

    for projectile in projectiles:
        projectile.show()
        projectile.move()
        for i in range(len(enemies) - 1, -1, -1):
            if projectile.hits(enemies[i]):
                del enemies[i]
                projectile.destroy()

    for enemy_projectile in enemy_projectiles:
        enemy_projectile.show()
        enemy_projectile.move()
        if enemy_projectile.hits(spaceship):
            print("Game Over")
            running = False

    for enemy in enemies:
        enemy.show()
        enemy.move()
        if random.random() < 0.001:
            enemy_projectiles.append(Projectile(enemy.x, enemy.y, True))

    projectiles = [p for p in projectiles if not p.to_delete]
    enemy_projectiles = [p for p in enemy_projectiles if not p.to_delete]

    pygame.display.flip()

pygame.quit()

File: space_invaders.js
// Global variables
let spaceship;
let enemies = [];
let projectiles = [];
let enemyProjectiles = [];
let gameOver = false;

function setup() {
  createCanvas(800, 600);
  startGame();
}

function draw() {
  background(0);
  spaceship.show();
  spaceship.move();

  if (!gameOver) {
    for (let projectile of projectiles) {
      projectile.show();
      projectile.move();
      for (let i = enemies.length - 1; i >= 0; i--) {
        if (projectile.hits(enemies[i])) {
          enemies.splice(i, 1);
          projectile.destroy();
        }
      }
    }

    for (let enemyProjectile of enemyProjectiles) {
      enemyProjectile.show();
      enemyProjectile.move();
      if (enemyProjectile.hits(spaceship)) {
        console.log("Game Over");
        gameOver = true;
      }
    }

    for (let enemy of enemies) {
      enemy.show();
      enemy.move();
      if (random() < 0.001) {
        enemyProjectiles.push(new Projectile(enemy.x, enemy.y, true));
      }
    }

    projectiles = projectiles.filter(p => !p.toDelete);
    enemyProjectiles = enemyProjectiles.filter(p => !p.toDelete);
  } else {
    textSize(32);
    fill(255);
    textAlign(CENTER, CENTER);
    text("Game Over", width / 2, height / 2);
    text("Press SPACE to restart", width / 2, height / 2 + 40);
  }
}

function startGame() {
  spaceship = new Spaceship();
  enemies = [];
  projectiles = [];
  enemyProjectiles = [];
  gameOver = false;

  for (let i = 0; i < 10; i++) {
    enemies.push(new Enemy(i * 80 + 80, 60));
  }
}

function keyPressed() {
  if (gameOver && key === ' ') {
    startGame();
  } else if (!gameOver) {
    if (key === ' ') {
      let projectile = new Projectile(spaceship.x, height - 20);
      projectiles.push(projectile);
    } else if (keyCode === RIGHT_ARROW) {
      spaceship.setDir(1);
    } else if (keyCode === LEFT_ARROW) {
      spaceship.setDir(-1);
    }
  }
}

class Spaceship {
  constructor() {
    this.x = width / 2;
    this.xdir = 0;
  }

  show() {
    fill(255);
    rectMode(CENTER);
    rect(this.x, height - 20, 20, 60);
  }

  move() {
    this.x += this.xdir * 5;
  }

  setDir(dir) {
    this.xdir = dir;
  }
}

class Enemy {
  constructor(x, y) {
    this.x = x;
    this.y = y;
    this.xdir = 1;
    this.speed = 2;
    this.amplitude = 50;
    this.frequency = 0.02;
    this.offset = random(0, 100);
  }

  show() {
    fill(255, 0, 0);
    ellipse(this.x, this.y, 60, 60);
  }

  move() {
    this.x += this.xdir * this.speed;
    this.y += sin((frameCount + this.offset) * this.frequency) * this.amplitude * this.frequency;
    if (this.x > width || this.x < 0) {
      this.xdir *= -1;
      this.y += 20;
    }
  }
}

class Projectile {
  constructor(x, y, isEnemy = false) {
    this.x = x;
    this.y = y;
    this.toDelete = false;
    this.isEnemy = isEnemy;
  }

  show() {
    fill(this.isEnemy ? 255 : 0, this.isEnemy ? 0 : 255, 0);
    ellipse(this.x, this.y, 8, 8);
  }

  move() {
    this.y += this.isEnemy ? 5 : -5;
  }

  hits(target) {
    let d = dist(this.x, this.y, target.x, target.y);
    return d < 30;
  }

  destroy() {
    this.toDelete = true;
  }
}

function keyReleased() {
  if (key !== ' ') {
    spaceship.setDir(0);
  }
}